Add Category model

dmarmul opened the pull request:
Added Category model and implemented the missing functionality.
Added Set<Category> categories to Book model.
Added Pagination, Sorting, and Swagger to all needed controllers

dmarmul committed:
Added Category model and implemented the missing functionality.
Added Set<Category> categories to Book model.
Added Pagination, Sorting, and Swagger to all needed controllers

Elena-Bruyako commented on src/main/java/org/example/bookshop/dto/CategoryDto.java:
```diff
@@ -0,0 +1,16 @@
+@ToString
+public class CategoryDto {
+    private Long id;
+    @NotBlank
```
```suggestion
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/BookServiceImpl.java:
```diff
@@ -56,4 +66,20 @@ public BookDto update(CreateBookRequestDto requestDto, Long id) {
+
+    private void setCategories(CreateBookRequestDto requestDto, Book book) {
+        book.setCategories(requestDto.getCategories().stream()
+                .map(c -> categoryRepository.findById(c.getId()).orElseThrow(()
```
rename c

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+    public CategoryDto update(CategoryDto categoryDto, Long id) {
+        Category category = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find category by id: " + id));
+        category.setName(categoryDto.getName());
```
use mapper to update category

liliia-ponomarenko replied:
not fixed

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
fixed some mistakes

liliia-ponomarenko commented on src/main/java/org/example/bookshop/controller/CategoryController.java:
```diff
@@ -0,0 +1,76 @@
+    @PostMapping
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Add category",
+            description = "Add a new category in db. Name can't be empty")
```
Use ResponseStatus.CREATED

liliia-ponomarenko commented on src/main/java/org/example/bookshop/controller/CategoryController.java:
```diff
@@ -0,0 +1,76 @@
+
+    @DeleteMapping("/{id}")
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Delete category", description = "Delete category by id")
```
ResponseStatus.NO_CONTENT

liliia-ponomarenko commented on src/main/java/org/example/bookshop/mapper/BookMapper.java:
```diff
@@ -2,13 +2,23 @@
+    BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
+
+    @AfterMapping
+    default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
+
+    }
```
Let's add realization ;)

liliia-ponomarenko commented on src/main/java/org/example/bookshop/dto/CreateBookRequestDto.java:
```diff
@@ -19,6 +20,7 @@ public class CreateBookRequestDto {
     @NotNull
     @Min(0)
     private BigDecimal price;
+    private Set<CategoryDto> categories;
```
You can NotEmpty validation

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/BookServiceImpl.java:
```diff
@@ -56,4 +66,21 @@ public BookDto update(CreateBookRequestDto requestDto, Long id) {
+                .toList();
+    }
+
+    private void setCategories(CreateBookRequestDto requestDto, Book book) {
+        book.setCategories(requestDto.getCategories().stream()
+                .map(categoryDto -> categoryRepository.findById(categoryDto.getId()).orElseThrow(()
+                        -> new EntityNotFoundException(
+                                "Can't find category by id: " + categoryDto.getId())
+                ))
+                .collect(Collectors.toSet()));
+    }
```
Try to use such logic in mapper 
```
  @AfterMapping
    default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
        bookDto.setCategoryIds(book.getCategories()
                .stream()
                .map(Category::getId)
                .toList()
        );
    }
```

liliia-ponomarenko commented on src/main/java/org/example/bookshop/service/impl/BookServiceImpl.java:
```diff
@@ -48,6 +57,7 @@ public BookDto update(CreateBookRequestDto requestDto, Long id) {
         book.setPrice(requestDto.getPrice());
         book.setDescription(requestDto.getDescription());
         book.setCoverImage(requestDto.getCoverImage());
+        setCategories(requestDto, book);
```
Use mapper for mapping Book

liliia-ponomarenko reviewed (changes requested):
Good job! Let's improve your solution ;)

dmarmul committed:
Added ResponseStatus for save and delete endpoints. Used mapper for saving and updating entities. Implemented setCategoryIds method in BookMapper.

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/BookServiceImpl.java:
```diff
@@ -3,51 +3,51 @@
 
     @Override
-    public List<BookDto> findAll(Pageable pageable) {
+    @Transactional
```
remove annotation

dmarmul replied:
but there is LazyInitializationException without it annotation, because of I need to return categories.
I don't need to use FetchType.EAGER, so how can I load categories, by @Query annotation?

JJJazl replied:
Not fixed
Yea, you can use custom select query with `@Query` annotation or `@EntityGraph(attributePaths = "categories")`

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/BookServiceImpl.java:
```diff
@@ -3,51 +3,51 @@
     }
 
     @Override
+    @Transactional
```
remove annotation

JJJazl replied:
Not fixed
See my comment about `@Query` and `@EntityGraph` annotations

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/BookServiceImpl.java:
```diff
@@ -3,51 +3,51 @@
-        book.setPrice(requestDto.getPrice());
-        book.setDescription(requestDto.getDescription());
-        book.setCoverImage(requestDto.getCoverImage());
+        Book book = bookMapper.toModel(requestDto);
```
```suggestion
        Book book = bookRepository.findById(id)
                .orElseThrow(()...
        bookMapper.updateBookFromDto(bookDto, book);
        return bookMapper.toDto(bookRepository.save(book));
```

Elena-Bruyako commented on src/main/java/org/example/bookshop/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public CategoryDto update(CategoryDto categoryDto, Long id) {
+        categoryDto.setId(id);
+        return categoryMapper.toDto(categoryRepository.save(
+                categoryMapper.toModel(categoryDto)));
```
```suggestion
        Category category = categoryRepository.findById(id)
                .orElseThrow(()....
        categoryMapper.updateCategoryFromDto(categoryDto, category);
        return categoryMapper.toDto(categoryRepository.save(category));
```

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
Added updateEntitiesFromDto in mappers,
used it in serviceImpls classes in update methods.
I don't know how Can I removed Transactional annotations because of LazyInitializationException. I don't need to use FetchType.EAGER, so how can I load categories, by @query annotation?

JJJazl commented on src/main/java/org/example/bookshop/mapper/BookMapper.java:
```diff
@@ -1,14 +1,32 @@
 
 @Mapper(config = MapperConfig.class)
 public interface BookMapper {
     BookDto toDto(Book book);
```
```suggestion
    @Mapping(target = "categoryIds", ignore = true)
    BookDto toDto(Book book);
```

JJJazl commented on src/main/resources/db/changelog/changes/13-create-categories-table.yaml:
```diff
@@ -0,0 +1,30 @@
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
It's better to have explicit names for constraints like: primary/foreign key, unique

dmarmul committed:
Used EntityGraph in repository instead of Transactional in serviceImpl class. Added primaryKeyName when create categories table.
Added categoryIds ignore annotation in BookMapper, because of there is AfterMapping for categoryIds already exist.

ahoienko reviewed (approved):
Good job!
